planner/core/rule_aggregation_elimination.go

```diff
@@ -155,3 +272,174 @@ func (a *aggregationEliminator) optimize(p LogicalPlan) (LogicalPlan, error) {
+// If the substitution happened, it returns true and the substituted expressions, otherwise it returns false and original exprs.
+func exprSubstitute(ctx sessionctx.Context, exprMap map[string]expression.Expression, exprs []expression.Expression) (bool, []expression.Expression) {
+	var newExprs []expression.Expression
+	replaced := false
+	for idx, expr := range exprs {
+		if e, ok := exprMap[string(expr.HashCode(ctx.GetSessionVars().StmtCtx))]; ok {
+			replaced = true
+			if newExprs == nil {
+				newExprs = make([]expression.Expression, len(exprs))
+				copy(newExprs, exprs)
+			}
+			newExprs[idx] = e
+			continue
+		}
+		sf, ok := expr.(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		if subReplaced, subExprs := exprSubstitute(ctx, exprMap, sf.GetArgs()); subReplaced {
+			replaced = true
+			newSf := expression.NewFunctionInternal(ctx, sf.FuncName.L, sf.GetType(), subExprs...)
+			if newExprs == nil {
+				newExprs = make([]expression.Expression, len(exprs))
+				copy(newExprs, exprs)
+			}
+			newExprs[idx] = newSf
+		}
+	}
+	if replaced {
+		return true, newExprs
+	}
+	return false, exprs
+}
+
+// aggSubstitute tries to substitutes expressions occurred in given `exprs` with the expression argument of aggregated functions defined in `aggMap`,
+// parameter `predicate` can be used to check each of the aggregated function.
+// This function returns:
+//  * true / substituted exprs if the substitution happens.
+//  * false / original exprs if substitution doesn't happen.
+//  * false / nil if predicate is not nil and it failed.
+func aggSubstitute(
+	ctx sessionctx.Context,
+	aggMap map[string]*aggregation.AggFuncDesc,
+	exprs []expression.Expression,
+	predicate func(fun *aggregation.AggFuncDesc) bool,
+) (bool, []expression.Expression) {
+	var newExprs []expression.Expression
+	replaced := false
+	for idx, expr := range exprs {
+		if fun, ok := aggMap[string(expr.HashCode(ctx.GetSessionVars().StmtCtx))]; ok {
+			// expr is a column
+			if predicate != nil && !predicate(fun) {
+				return false, nil
+			}
+			replaced = true
+			if newExprs == nil {
+				newExprs = make([]expression.Expression, len(exprs))
+				copy(newExprs, exprs)
+			}
+			newExprs[idx] = fun.Args[0]
+			continue
+		}
+		sf, ok := expr.(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		subReplaced, subExprs := aggSubstitute(ctx, aggMap, sf.GetArgs(), predicate)
+		if subExprs == nil {
+			return false, nil
+		}
+		if subReplaced {
+			replaced = true
+			newSf := expression.NewFunctionInternal(ctx, sf.FuncName.L, sf.GetType(), subExprs...)
+			if newExprs == nil {
+				newExprs = make([]expression.Expression, len(exprs))
+				copy(newExprs, exprs)
+			}
+			newExprs[idx] = newSf
+		}
+	}
+	if replaced {
+		return true, newExprs
+	}
+	return false, exprs
+}
+
+func deepContains(ctx sessionctx.Context, exprs []expression.Expression, expr expression.Expression) bool {
```

Comment:
How about renaming it to `EqualContains` and move it to be with `Contains` in expression/util.go?

Reply:
Addressed